### PR TITLE
Implement alpha to coverage (A2C) support.

### PR DESCRIPTION
**Pull request by pcwalton**
[Alpha to coverage] (A2C) replaces alpha blending with a hardware-specific multisample coverage mask when multisample antialiasing is in use. It's a simple form of [order-independent transparency] that relies on MSAA. ["Anti-aliased Alpha Test: The Esoteric Alpha To Coverage"] is a good summary of the motivation for and best practices relating to A2C.

This commit implements alpha to coverage support as a new variant for `AlphaMode`. You can supply `AlphaMode::AlphaToCoverage` as the `alpha_mode` field in `StandardMaterial` to use it. When in use, the standard material shader automatically applies the texture filtering method from ["Anti-aliased Alpha Test: The Esoteric Alpha To Coverage"]. Objects with alpha-to-coverage materials are binned in the opaque pass, as they're fully order-independent.

The `transparency_3d` example has been updated to feature an object with alpha to coverage. Happily, the example was already using MSAA.

This is part of #2223, as far as I can tell.

[Alpha to coverage]: https://en.wikipedia.org/wiki/Alpha_to_coverage

[order-independent transparency]: https://en.wikipedia.org/wiki/Order-independent_transparency

["Anti-aliased Alpha Test: The Esoteric Alpha To Coverage"]: https://bgolus.medium.com/anti-aliased-alpha-test-the-esoteric-alpha-to-coverage-8b177335ae4f

---

### Added

* The `AlphaMode` enum now supports `AlphaToCoverage`, to provide limited order-independent transparency when multisample antialiasing is in use.

### CODE REVIEW

File: crates/bevy_pbr/src/pbr_material.rs [hunk 1]
```diff
@@ -648,6 +648,7 @@ bitflags::bitflags! {
         const ALPHA_MODE_PREMULTIPLIED   = 3 << Self::ALPHA_MODE_SHIFT_BITS;                          //
         const ALPHA_MODE_ADD             = 4 << Self::ALPHA_MODE_SHIFT_BITS;                          //   Right now only values 0–5 are used, which still gives
         const ALPHA_MODE_MULTIPLY        = 5 << Self::ALPHA_MODE_SHIFT_BITS;                          // ← us "room" for two more modes without adding more bits
+        const ALPHA_MODE_A2C             = 6 << Self::ALPHA_MODE_SHIFT_BITS;
```
Comment (superdump) [inline crates/bevy_pbr/src/pbr_material.rs:651]:
```suggestion
        const ALPHA_MODE_ALPHA_TO_COVERAGE             = 6 << Self::ALPHA_MODE_SHIFT_BITS;
```

File: crates/bevy_pbr/src/pbr_material.rs [hunk 2]
```diff
@@ -783,6 +784,7 @@ impl AsBindGroupShaderType<StandardMaterialUniform> for StandardMaterial {
             AlphaMode::Premultiplied => flags |= StandardMaterialFlags::ALPHA_MODE_PREMULTIPLIED,
             AlphaMode::Add => flags |= StandardMaterialFlags::ALPHA_MODE_ADD,
             AlphaMode::Multiply => flags |= StandardMaterialFlags::ALPHA_MODE_MULTIPLY,
+            AlphaMode::AlphaToCoverage => flags |= StandardMaterialFlags::ALPHA_MODE_A2C,
```
Comment (superdump) [inline crates/bevy_pbr/src/pbr_material.rs:787]:
```suggestion
            AlphaMode::AlphaToCoverage => flags |= StandardMaterialFlags::ALPHA_MODE_ALPHA_TO_COVERAGE,
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 3]
```diff
@@ -1056,6 +1056,7 @@ bitflags::bitflags! {
         const BLEND_PREMULTIPLIED_ALPHA         = 1 << Self::BLEND_SHIFT_BITS;                   //
         const BLEND_MULTIPLY                    = 2 << Self::BLEND_SHIFT_BITS;                   // ← We still have room for one more value without adding more bits
         const BLEND_ALPHA                       = 3 << Self::BLEND_SHIFT_BITS;
+        const BLEND_A2C                         = 4 << Self::BLEND_SHIFT_BITS;
```
Comment (superdump) [inline crates/bevy_pbr/src/render/mesh.rs:1059]:
```suggestion
        const BLEND_ALPHA_TO_COVERAGE           = 4 << Self::BLEND_SHIFT_BITS;
```

Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: Made all these renamings.

File: crates/bevy_pbr/src/render/pbr_types.wgsl [hunk 4]
```diff
@@ -51,6 +51,7 @@ const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BLEND: u32               = 1073741824u;
 const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_PREMULTIPLIED: u32       = 1610612736u; // (3u32 << 29)
 const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_ADD: u32                 = 2147483648u; // (4u32 << 29)
 const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_MULTIPLY: u32            = 2684354560u; // (5u32 << 29)
+const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_A2C: u32                 = 3221225472u; // (6u32 << 29)
```
Comment (superdump) [inline crates/bevy_pbr/src/render/pbr_types.wgsl]:
```suggestion
const STANDARD_MATERIAL_FLAGS_ALPHA_MODE_ALPHA_TO_COVERAGE: u32   = 3221225472u; // (6u32 << 29)
```

**Review by superdump — APPROVED** [review]
The MIP-level scaling, which could be a preprocessing thing, can wait I suppose.

The article also mentions an issue with Unity's cascaded shadow mapping that I'm wondering if we are also affected by and need to address. But that can also be done in a follow-up. I would appreciate confirmation of whether it is an issue for us though. :)

**Comment by pcwalton** [comment]
@superdump I've changed the logic to fall back to alpha mask with a cutoff of 0.5 if MSAA is off.

As for the Unity bug: Translated to Bevy concepts, the Unity issue is that if you use `ExtendedMaterial<StandardMaterial, ...>` with alpha to coverage on, shadows won't show up because Unity's shadow mapping has a fragment shader that leaves the alpha value at 0.0, and alpha to coverage is enabled during the shadow pass. Bevy is unaffected by this for two reasons. First, Bevy doesn't enable alpha to coverage during the shadow pass. Second, Bevy disables the fragment shader entirely during the shadow mapping pass. [Section 26.4 of the Vulkan spec](https://registry.khronos.org/vulkan/specs/1.3/html/vkspec.html#fragops-covg) requires that, when the fragment shader is disabled, alpha to coverage doesn't modify the coverage mask, which prevents the Unity bug.

**Review by superdump — APPROVED** [review]

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl [hunk 5]
```diff
@@ -107,9 +107,23 @@ fn pbr_input_from_standard_material(
...
+    let alpha_mode = pbr_bindings::material.flags &
+        pbr_types::STANDARD_MATERIAL_FLAGS_ALPHA_MODE_RESERVED_BITS;
+    if alpha_mode == pbr_types::STANDARD_MATERIAL_FLAGS_ALPHA_MODE_ALPHA_TO_COVERAGE {
+        pbr_input.material.base_color.a = (pbr_input.material.base_color.a -
+                pbr_bindings::material.alpha_cutoff) /
+                max(fwidth(pbr_input.material.base_color.a), 0.0001) + 0.5;
+    }
+#endif // ALPHA_TO_COVERAGE
+
+#endif // MESHLET_MESH_MATERIAL_PASS
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_fragment.wgsl:126]: Is a2c not compatible with meshlets?

**Comment by pcwalton** [comment]
@IceSentry The shadow behavior for A2C now matches blend.

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl (hunk 5 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_fragment.wgsl:126]: I don't know enough about meshlets to know. I know that Nanite is incompatible with transparency. I'd rather not touch the meshlet code as it's being heavily worked on.

Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_fragment.wgsl]: I moved the code out of the meshlet block.

**Comment by IceSentry** [comment]
Alright, just one final tiny nitpick, in the example, on the system that updates the alpha of all meshes. There should be a short description of what happens to a2c meshes.

**Comment by pcwalton** [comment]
@IceSentry Done.

**Review by IceSentry — APPROVED** [review]

**Pull request merged**